### PR TITLE
Do not forward channel twice

### LINKED ISSUE #124: Do not forward an EPICS channel more than once
When a command attempts to add the same channel a second time, the forwarder should refuse.

### PR DESCRIPTION
**Pull request by dominikwerder**
### Description of work

Before adding a channel for forwarding, check whether the channel is already being forwarded.

If channel is already forwarded, do not add it a second time.

### Issue

Closes #124 

### Acceptance Criteria

Check that adding a channel twice only adds it one. There is a corresponding system test for this

### Unit Tests

Added a couple of unit tests for some of the changes.
Created a separate issue for adding unit tests for the Stream class (#146)

### Other

System test added.

---

## Code Review (To be filled in by the reviewer only)

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/Streams.cpp [hunk 1]
```diff
@@ -59,23 +52,29 @@ void Streams::check_stream_status() {
...
+/// Return true is the given channel name is already forwarded
+bool Streams::hasChannelName(std::string const &channel_name) {
+  for (auto const &x : streams) {
+    if (x->channel_info().channel_name == channel_name) {
+      return true;
+    }
+  }
+  return false;
+}
+}
```
Comment (mattclarke) [inline src/Streams.cpp]: Could this potentially use std::find_if?
Though this way is more readable... <shrug>

**Comment by dominikwerder** [comment]
In the current PR, you can still add more than one converter and forward to more than one cluster.
Only if the channel is already forwarded you would have to remove it first, and then add it again with the new set of converters.
But of course we can change this PR to refuse only if channel and kafka URI and converter is the same, what do you think?

**Comment by mattclarke** [comment]
That sounds reasonable to me based on my limited understanding of the requirements.

**Comment by matthew-d-jones** [comment]
Refusing only if the channel, kafka URI and converter are the same is the behaviour I would expect. 
Otherwise the Forwarder dictates that it's configuration changes are all sent from a single producer which is also keeping track of the full configuration.

File: src/Stream.cpp [hunk 2]
```diff
@@ -47,6 +47,20 @@ nlohmann::json ConversionPath::status_json() const {
   return Document;
 }
 
+std::string ConversionPath::getKafkaTopicName() const {
+  if (kafka_output) {
+    return kafka_output->topic_name();
+  }
+  return "nullptr";
```
Comment (rerpha) [inline src/Stream.cpp]: is `return "nullptr"` the right thing to do here? could we not throw an exception or log?

File: src/Stream.cpp [hunk 3]
```diff
@@ -47,6 +47,20 @@ nlohmann::json ConversionPath::status_json() const {
...
+}
+
+std::string ConversionPath::getSchemaName() const {
+  if (converter) {
+    converter->schema_name();
+  }
+  return "nullptr";
```
Comment (rerpha) [inline src/Stream.cpp]: is return "nullptr" the right thing to do here? could we not throw an exception or log?

File: src/Forwarder.cpp [hunk 4]
```diff
@@ -348,43 +348,53 @@ void Forwarder::addMapping(StreamSettings const &StreamInfo) {
...
-  for (auto &Converter : StreamInfo.Converters) {
-    pushConverterToStream(Converter, Stream);
-  }
 }
 
 template <typename T>
-std::shared_ptr<T> Forwarder::addStream(ChannelInfo &ChannelInfo) {
+std::shared_ptr<Stream> Forwarder::findOrAddStream(ChannelInfo &ChannelInfo) {
+  std::shared_ptr<Stream> FoundStream =
+      streams.getStreamByChannelName(ChannelInfo.channel_name);
```
Comment (rerpha) [inline src/Forwarder.cpp]: maybe add some behaviour for if a stream is not found

File: src/Streams.cpp [hunk 5]
```diff
@@ -67,4 +67,15 @@ std::shared_ptr<Stream> Streams::back() {
 const std::vector<std::shared_ptr<Stream>> &Streams::get_streams() {
   return streams;
 }
+
+/// Return true is the given channel name is already forwarded
```
Comment (mattclarke) [inline src/Streams.cpp]: This comment is out of date!

**Pull request merged**